### PR TITLE
Metrics config.

### PR DESCRIPTION
**Pull request by rasapala**

### CODE REVIEW

File: src/metric_config.cpp [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include "metric_config.hpp"
+
+#include <algorithm>
+#include <filesystem>
```
Comment (dkalinowski) [inline src/metric_config.cpp:19]: What is it used for?

File: src/metric_config.cpp [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+#include <limits>
```
Comment (dkalinowski) [inline src/metric_config.cpp:20]: What is it used for?

File: src/metric_config.cpp [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+#include <regex>
+#include <set>
+#include <sstream>
```
Comment (dkalinowski) [inline src/metric_config.cpp:23]: What is it used for?

File: src/metric_config.cpp [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include "metric_config.hpp"
+
+#include <algorithm>
```
Comment (dkalinowski) [inline src/metric_config.cpp:18]: What is it used for?

File: src/metric_config.cpp (hunk 1 above)
Comment (rasapala) [reply src/metric_config.cpp:19]: Removed.

File: src/metric_config.cpp (hunk 2 above)
Comment (rasapala) [reply src/metric_config.cpp:20]: Removed.

File: src/metric_config.cpp [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+#include <filesystem>
+#include <limits>
+#include <regex>
+#include <set>
```
[the remark this answers is not in the document]
Comment (rasapala) [reply src/metric_config.cpp:22]: Removed.

File: src/metric_config.cpp (hunk 4 above)
Comment (rasapala) [reply src/metric_config.cpp:18]: Removed.

**Review by dkalinowski — APPROVED** [review]

File: src/metric_config.hpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    Status parseMetricsConfig(const rapidjson::Value& v);
+    bool validateEndpointPath(std::string endpoint);
+
+    void setAllMetricsTo(bool enabled);
+
+    MetricConfig() {
+        metricsEnabled = false;
+        endpointsPath = "/metrics";
+
+        setAllMetricsTo(false);
```
Comment (atobiszei) [inline src/metric_config.hpp:78]:
```suggestion
        setAllMetricsTo(metricsEnabled);
```

File: src/metric_config.hpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    MetricConfig(bool enabled) {
+        metricsEnabled = enabled;
+        endpointsPath = "/metrics";
+
+        setAllMetricsTo(enabled);
```
Comment (atobiszei) [inline src/metric_config.hpp:85]:
```suggestion
        setAllMetricsTo(metricsEnabled);
```

File: src/test/metric_config_test.cpp [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+                "nireq": 100,
+                "shape": {"b": "(1,10) "}
+            }
+        }
+    ],
+    "monitoring":
+        {
+            "metrics":
+            {
+                "enable" : true,
```
Comment (atobiszei) [inline src/test/metric_config_test.cpp:113]: Add few tests for negative config. Eg:
1) add not required field
2) "enable": 1
3) "enable" : {}
4) endpoint_path : {"a":3}
5) enable" :{}

File: src/test/metric_config_test.cpp [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+    "monitoring":
+        {
+            "metrics":
+            {
+                "enable" : true
+            }
+        }
+})";
+
+static const char* modelMetricsBadEndpoint = R"(
```
Comment (atobiszei) [inline src/test/metric_config_test.cpp:95]: Minor: In my opinion test is more readable if the config is just above the test that uses that config

File: src/test/metric_config_test.cpp (hunk 8 above)
Comment (atobiszei) [reply src/test/metric_config_test.cpp:113]: Those probably should be in schema_test.cpp

File: src/test/metric_config_test.cpp (hunk 9 above)
Comment (rasapala) [reply src/test/metric_config_test.cpp:95]: Done.

File: src/test/metric_config_test.cpp (hunk 8 above)
Comment (rasapala) [reply src/test/metric_config_test.cpp:113]: Done.

File: src/test/metric_config_test.cpp [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+})";
+
+TEST_F(MetricsConfigTest, MetricsNegative2) {
+    SetUpConfig(modelMetricsNegative2);
+    std::filesystem::copy("/ovms/src/test/dummy", modelPath, std::filesystem::copy_options::recursive);
+    createConfigFileWithContent(ovmsConfig, configFilePath);
+    ConstructorEnabledModelManager manager;
+
+    auto status = manager.loadConfig(configFilePath);
+    ASSERT_FALSE(status.ok());
```
Comment (atobiszei) [inline src/test/metric_config_test.cpp:259]: Minor:
when you test for status it's usually better to test for specific status code to know when you change the behavior. Additionally it's good to have sth like
ASSERT_TRUE(status.ok()) << status.string(); to have quicker debug in case of errors.

**Review by atobiszei — APPROVED** [review]
